hs/src/Reach/Simulator/Server.hs

```diff
@@ -0,0 +1,154 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
+{-# OPTIONS_GHC -Wno-deriving-defaults #-}
+{-# OPTIONS_GHC -Wno-missing-methods #-}
+
+module Reach.Simulator.Server where
+
+import Reach.AST.LL
+import Control.Monad.Reader
+import qualified Reach.Simulator.Core as C
+import Control.Concurrent.STM
+import Data.Default.Class
+import Data.Text.Lazy (Text)
+import Network.Wai.Middleware.RequestLogger
+import Web.Scotty.Trans
+import qualified Data.Map.Strict as M
+-- import GHC.Generics
+-- import Data.Aeson (FromJSON, ToJSON)
+
+instance Default Session where
+  def = initSession
+
+newtype WebM a = WebM { runWebM :: ReaderT (TVar Session) IO a }
+  deriving (Applicative, Functor, Monad, MonadIO, MonadReader (TVar Session))
+
+webM :: MonadTrans t => WebM a -> t WebM a
+webM = lift
+
+gets :: (Session -> b) -> WebM b
+gets f = ask >>= liftIO . readTVarIO >>= return . f
+
+modify :: (Session -> Session) -> WebM ()
+modify f = ask >>= liftIO . atomically . flip modifyTVar' f
+
+type StateId = Int
+type ActionId = Int
+
+portNumber :: Int
+portNumber = 3000
+
+-- state
+data Session = Session
+  {  e_states_actions :: M.Map StateId [ActionId]
+  ,  e_nsid :: Int -- next state id
+  ,  e_naid :: Int -- next action id
+  ,  e_res :: C.DLVal
+  ,  e_ids_actions :: M.Map ActionId C.Action
+  ,  e_states_ks :: M.Map StateId C.PartState
+  }
+
+initSession :: Session
+initSession = Session
+  { e_states_actions = M.empty
+  , e_nsid = 0
+  , e_naid = 0
+  , e_res = C.V_Null
+  , e_ids_actions = M.empty
+  , e_states_ks = M.empty
+  }
+
+initProgSim :: LLProg -> WebM C.PartState
+initProgSim ll = do
+  let initSt = C.initState
+  ps <- return $ C.initApp ll initSt
+  processNewState ps
+
+processNewState :: C.PartState -> WebM C.PartState
+processNewState ps = do
+  sid <- gets e_nsid
+  _ <- case ps of
+    C.PS_Done _ _ -> do
+      _ <- return $ putStrLn "EVAL DONE"
+      registerAction sid C.A_None
+    C.PS_Suspend a _ _ -> registerAction sid a
```

Comment:
There may be a modified state as well

Reply:
@jeapostrophe yes, we save that later on